### ISSUE TITLE
Consider refactoring integration tests into their own module

### ISSUE THREAD

**Issue by rnc** [release-engineering/pom-manipulation-ext#177]
Ensure they can be run in either ext mode or in CLI mode.

**Comment by jdcasey** [comment]
We could put the tests themselves (project directory content) in a separate module, then provide test "drivers" that reference these projects and run the builds...for the ext, it'd probably still use invoker, but the cli would probably use normal junit tests.

Is that what you have in mind?

**Comment by rnc** [comment]
Yes pretty much -  a integration-test module, with perhaps invoker for ext tests and some other form of driver for the CLI - ideally all based on the same set of tests. Although we might end up needed exclusion/inclusion in case some tests are not applicable.

**Comment by rnc** [comment]
I think the createSession in Cli needs to support setting a local repository (and plugin repository?) .... i.e. equivalent to the Maven -s command.

**Comment by VaclavDedik** [comment]
So I am currently using Junit to execute the tests in the cli mode. I am reusing basically everything in the "it" directory (test.properties, pom.xml, verify.groovy), there is a problem though. When I execute the PME in cli mode, it doesn't do "mvn install" as maven-invoker-plugin does. So when the "verify.groovy" is executed, the test fails because some of the integration tests check for jar (and pom).

One solution I can think of is to remove these checks from the tests. In my opinion this wouldn't be such a big deal because I think the tests should test if PME works correctly, not if "mvn install" works correctly.

Another solution would be to run only some tests in the cli mode.

What do you think?

**Comment by jdcasey** [comment]
I think I'd prefer to exclude those tests from the CLI suite's execution, IF the tests are really checking something related to the build itself (for instance, that maven is actually using the changed POMs). Otherwise, I'd say modifying the verify.groovy scripts would be fine.

If we do run into tests that have important verifications but *also* really need to verify output from a maven build, I'd suggest cloning them and creating a version that only verifies the maven install output (which would be excluded from CLI tests), and a version that only verifies the other stuff...basically decomposing the tests. Tests like that are a sort of bad code smell anyway, in that they wouldn't really be testing a single feature/behavior.

**Comment by rnc** [comment]
From taking a quick look at the groovy scripts they mostly seem to be doing this kind of thing
```
def jar = new File(basedir, "target/${pom.artifactId.text()}-${pom.version.text()}.jar" )
System.out.println( "Checking for jar: ${jar.getAbsolutePath()}")
assert jar.exists()

def repodir = new File('@localRepositoryUrl@', "${pom.groupId.text().replace('.', '/')}/${pom.artifactId.text()}/${pom.version.text()}" )
def repojar = new File( repodir, "${pom.artifactId.text()}-${pom.version.text()}.jar" )
System.out.println( "Checking for installed jar: ${repojar.getAbsolutePath()}")
assert repojar.exists()

def repopom = new File( repodir, "${pom.artifactId.text()}-${pom.version.text()}.pom" )
System.out.println( "Checking for installed pom: ${repopom.getAbsolutePath()}")
assert repopom.exists()
integration-test/src/it/repo-removal/verify.groovy (END)
```

So essentially verifying the invoked maven project and the installed jars. Is it worth invoking Maven manually and then running the groovy script ?

**Comment by VaclavDedik** [comment]
@rnc That sounds like a good idea. I tried it and it works quite well. I am using this command to run maven:

    mvn install -Dmaven.repo.local=${maven.build.directory}/local-repo -s ../interpolated-settings.xml

Not sure if there is a better way.

Anyway, I think I'll do it this way, we can change it later if we want.

**Comment by VaclavDedik** [comment]
I am almost done. There are two problems however:

1. The -s option for the pom-manipulation-cli is necessary. There are 32 test errors that are caused by not having this option.
2. I think that it would be a good idea to update JUnit to version 4.11. I am using parameterized tests, the problem is that in JUnit 4.10, there is no way to name each run of the test with different parameter. This means that the test runs are only assigned a number, so if there is a test failure/error, it's impossible to tell which integration test actually failed (if it's for example, simple-numeric or some other etc.). In JUnit 4.11, there is a way to name each test run, see http://stackoverflow.com/questions/650894/changing-names-of-parameterized-tests

**Comment by rnc** [comment]
@VaclavDedik +1 on upgrade (or even to 4.12).
Note some of the tests might be failed due to bugs in CLI - I think it needs ability to configure local repository as I mentioned offline.

**Comment by VaclavDedik** [comment]
@rnc Configure local repository: Yes, that's what I meant by the '-s' option, i.e. support for custom settings.xml file. I believe that all 32 test errors are caused by not being able to set local repository, because in all of those 32 test I get the following error:

```
[INFO] Maven-Manipulation-Extension 1.7-SNAPSHOT
[main] INFO  o.c.m.e.m.i.ProjectSourcesInjectingManipulator - Examining org.commonjava.maven.ext:basic-plugin-management-override:1.0-SNAPSHOT [pom=/home/vdedik/Projects/pom-manipulation-ext/integration-test/target/it-cli/remote-plugin-management-basic/pom.xml] to apply sources/metadata plugins.
[main] ERROR org.commonjava.maven.ext.manip.Cli - Unable to parse projects 
org.commonjava.maven.ext.manip.ManipulationException: Failed to resolve POM: org.commonjava.maven.ext:pluginMgmt1:pom:1.0
	at org.commonjava.maven.ext.manip.io.ModelIO.resolveRawModel(ModelIO.java:93) ~[pom-manipulation-cli.jar:1.7-SNAPSHOT]
	at org.commonjava.maven.ext.manip.io.ModelIO.getRemotePluginVersionOverrides(ModelIO.java:166) ~[pom-manipulation-cli.jar:1.7-SNAPSHOT]
	at org.commonjava.maven.ext.manip.impl.PluginManipulator.loadRemoteBOM(PluginManipulator.java:152) ~[pom-manipulation-cli.jar:1.7-SNAPSHOT]
	at org.commonjava.maven.ext.manip.impl.PluginManipulator.applyChanges(PluginManipulator.java:117) ~[pom-manipulation-cli.jar:1.7-SNAPSHOT]
	at org.commonjava.maven.ext.manip.ManipulationManager.applyManipulations(ManipulationManager.java:194) ~[pom-manipulation-cli.jar:1.7-SNAPSHOT]
	at org.commonjava.maven.ext.manip.ManipulationManager.scanAndApply(ManipulationManager.java:134) ~[pom-manipulation-cli.jar:1.7-SNAPSHOT]
	at org.commonjava.maven.ext.manip.Cli.run(Cli.java:156) [pom-manipulation-cli.jar:1.7-SNAPSHOT]
	at org.commonjava.maven.ext.manip.Cli.main(Cli.java:65) [pom-manipulation-cli.jar:1.7-SNAPSHOT]
```

**Comment by VaclavDedik** [comment]
This is getting interesting... so I implemented the '-s' option (with @jdcasey help, thanks!), but now there is another problem it seems.. when I run the remote-dependency-management-property test in cli mode with the '-s' option, the pom-manipulation-cli does what it is supposed to do, but at the very end the program hangs. The output looks like this:

```
java -jar ../../pom-manipulation-cli.jar -s /home/vdedik/Projects/pom-manipulation-ext/integration-test/target/it-cli/settings.xml -DdependencyManagement=org.commonjava.maven.ext\:depMgmt1\:1.0
[INFO] Maven-Manipulation-Extension 1.7-SNAPSHOT
[main] INFO  o.c.m.e.m.i.ProjectSourcesInjectingManipulator - Examining org.commonjava.maven.ext:basic-dependency-property:1.0-SNAPSHOT [pom=/home/vdedik/Projects/pom-manipulation-ext/integration-test/target/it-cli/remote-dependency-management-property/pom.xml] to apply sources/metadata plugins.
[main] INFO  o.c.m.e.m.impl.DependencyManipulator - Updating property version.junit with 4.1 
[main] INFO  o.c.m.ext.manip.ManipulationManager - Maven-Manipulation-Extension: Rewrite changed: [org.commonjava.maven.ext:basic-dependency-property:1.0-SNAPSHOT [pom=/home/vdedik/Projects/pom-manipulation-ext/integration-test/target/it-cli/remote-dependency-management-property/pom.xml]]
[main] INFO  o.c.maven.ext.manip.io.PomIO - org.commonjava.maven.ext:basic-dependency-property:1.0-SNAPSHOT [pom=/home/vdedik/Projects/pom-manipulation-ext/integration-test/target/it-cli/remote-dependency-management-property/pom.xml] modified! Rewriting.
[main] INFO  o.c.maven.ext.manip.io.PomIO - Rewriting: org.commonjava.maven.ext:basic-dependency-property:jar:1.0-SNAPSHOT in place of: org.commonjava.maven.ext:basic-dependency-property:jar:1.0-SNAPSHOT
       to POM: /home/vdedik/Projects/pom-manipulation-ext/integration-test/target/it-cli/remote-dependency-management-property/pom.xml
[main] INFO  o.c.m.ext.manip.ManipulationManager - Maven-Manipulation-Extension: Finished.
^C
```

What is interesting is that if I run it again (with target dir not deleted), it dosn't hang anymore. In fact it seems that it hangs every time I delete 'manipulator-cache' from target directory. Not sure if this is caused by addition of some new code or if this would happen either way. Gona have to look into it tomorrow.

**Comment by VaclavDedik** [comment]
So I finally managed to fix the hanging issue.. turned out someone forgot to shutdown ExecutorService in GalleyInfrastructure.

Anyway, I also finally managed to fix all errors (yay!), but, there are still a couple of failures (5 to be exact), that's because these have some quite complicated execution (i.e. more than just 'invoker.goals=install' in invoker.properties)... So I guess I'll try to fix as many of them as possible, if it turns out to be too complicated, I will just exclude them and those will have to be tested manually (i.e. write a seperate junit test).

This is current status:
```
Failed tests: 
  DefaultCliIntegrationTest.testIntegration:74 assert exceptionRaised == true
       |               |
       false           false
  DefaultCliIntegrationTest.testIntegration:74 assert dependency.version.text() == "1.0.12"
       |          |       |      |
       |          1.0.11  1.0.11 false
       ch.qos.logbacklogback-classic1.0.11
  DefaultCliIntegrationTest.testIntegration:74 assert dependency.version.text() == "4.10"
       |          |       |      |
       |          3.8.1   3.8.1  false
       junitjunit3.8.1test
  DefaultCliIntegrationTest.testIntegration:74 assert message == 1
       |       |
       0       false
  DefaultCliIntegrationTest.testIntegration:74 assert dependency.version.text() == "4.1"
       |          |       |      |
       |          3.8.2   3.8.2  false
       junitjunit3.8.2

Tests run: 57, Failures: 5, Errors: 0, Skipped: 0

```

**Comment by jdcasey** [comment]
this is looking great! Let me know if there's anything you need my help with.

**Comment by VaclavDedik** [comment]
Great, thanks, but I think I am done. Everything seems to be working on my laptop so I am going to create a PR. I managed to fix all failures (although the test code for it got quite complicated as I had to implement quite an extensive part of invoker plugin behavior).

**Comment by rnc** [comment]
@VaclavDedik - this is great, thanks :-)

**Comment by rnc** [comment]
Fixed by #185
